consensus/blake3pow/consensus.go

```diff
@@ -643,31 +648,282 @@ func (blake3pow *Blake3pow) Finalize(chain consensus.ChainHeaderReader, header *
+	utxosCreated, _ := rawdb.ReadPrunedUTXOKeys(chain.Database(), blockHeight)
+	if len(utxosCreated) == 0 {
+		// This should almost never happen, but we need to handle it
+		utxosCreated, _ = rawdb.ReadCreatedUTXOKeys(chain.Database(), blockHash)
+		logger.Infof("Reading non-pruned UTXOs for block %d", blockHeight)
+		for i, key := range utxosCreated {
+			if len(key) == rawdb.UtxoKeyWithDenominationLength {
+				if key[len(key)-1] > types.MaxTrimDenomination {
+					// Don't keep it if the denomination is not trimmed
+					// The keys are sorted in order of denomination, so we can break here
+					break
+				}
+				key[rawdb.PrunedUtxoKeyWithDenominationLength+len(rawdb.UtxoPrefix)-1] = key[len(key)-1] // place the denomination at the end of the pruned key (11th byte will become 9th byte)
+			}
+			// Reduce key size to 9 bytes and cut off the prefix
+			key = key[len(rawdb.UtxoPrefix) : rawdb.PrunedUtxoKeyWithDenominationLength+len(rawdb.UtxoPrefix)]
+			utxosCreated[i] = key
+		}
+	}
+
+	logger.Infof("UTXOs created in block %d: %d", blockHeight, len(utxosCreated))
+	if len(collidingKeys) > 0 {
+		logger.Infof("Colliding keys: %d", len(collidingKeys))
+		utxosCreated = append(utxosCreated, collidingKeys...)
+		sort.Slice(utxosCreated, func(i, j int) bool {
+			return utxosCreated[i][len(utxosCreated[i])-1] < utxosCreated[j][len(utxosCreated[j])-1]
+		})
+	}
+	newCollisions := make([][]byte, 0)
+	duplicateKeys := make(map[[36]byte]bool) // cannot use rawdb.UtxoKeyLength for map as it's not const
+	// Start by grabbing all the UTXOs created in the block (that are still in the UTXO set)
+	for _, key := range utxosCreated {
+		if len(key) != rawdb.PrunedUtxoKeyWithDenominationLength {
+			continue
+		}
+		if checkDenom {
+			if key[len(key)-1] != denomination {
+				if key[len(key)-1] > denomination {
+					break // The keys are stored in order of denomination, so we can stop checking here
+				} else {
+					continue
+				}
+			} else {
+				key = append(rawdb.UtxoPrefix, key...) // prepend the db prefix
+				key = key[:len(key)-1]                 // remove the denomination byte
+			}
+		}
+		// Check key in database
+		i := 0
+		it := chain.Database().NewIterator(key, nil)
+		for it.Next() {
+			data := it.Value()
+			if len(data) == 0 {
+				logger.Infof("Empty key found, denomination: %d", denomination)
+				continue
+			}
+			// Check if the key is a duplicate
+			if len(it.Key()) == rawdb.UtxoKeyLength {
+				key36 := [36]byte(it.Key())
+				if duplicateKeys[key36] {
+					continue
+				} else {
+					duplicateKeys[key36] = true
+				}
+			}
+			utxoProto := new(types.ProtoTxOut)
+			if err := proto.Unmarshal(data, utxoProto); err != nil {
+				logger.Errorf("Failed to unmarshal ProtoTxOut: %+v data: %+v key: %+v", err, data, key)
+				continue
```

Comment:
can we continue if any of this errors out? then we are not following the trim protocol right

Reply:
Asking if we should reject the block

Reply:
If len(data) = 0 it’s possible that the UTXO was spent. But if we can’t proto unmarshal the UTXO, then our node is broken, so we should probably reject.

Reply:
@gameofpointers now that we know that sometimes data can end up in the prefix... do you still want to reject the block here if it doesn't proto unmarshal to UTXO?